Draggable feedback positioning

timcreatedit opened the pull request:
Reopened after revert in #147658

Another test was added in the meantime to `draggable_test.dart` that didn't call `gesture.up()`. I added this call to the test and now all tests pass.

---
# Original Description (#145647):
We changed the coordinates used to position the `Draggable` feedback by transforming them into the `Overlay`s coordinate space. This has no influence on any untransformed `Overlay`, most Flutter apps should be not affected. 
This PR fixes the positioning of the feedback in transformed context (see #145639 for before video):

https://github.com/flutter/flutter/assets/42270125/df34e198-0667-453d-a27a-a79b2e2825a1

- fixes #145639 

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

timcreatedit committed:
use overlay coordinate space for draggable feedback

JesperBellenbaum committed:
use late

timcreatedit committed:
added test for non-transformed draggable

timcreatedit committed:
added tests for scaled and unscaled apps

timcreatedit committed:
added expected header to new test file and formatted it

timcreatedit committed:
finish gestures in tests

timcreatedit committed:
moved tests to correct file

timcreatedit committed:
finished gestures on tests in draggable_test.dart

timcreatedit committed:
added test for rotation

timcreatedit committed:
added test for dragAnchorStrategy

timcreatedit committed:
removed unneccessary import

timcreatedit committed:
released resources in remaining test in draggable_test.dart

dkwingsmt reviewed (approved):
LGTM!

justinmc commented on packages/flutter/test/widgets/draggable_test.dart:
```diff
@@ -3497,6 +3529,93 @@ void main() {
+
+    // Finish gesture to release resources.
+    await gesture.up();
+    await tester.pumpAndSettle();
```
Nit: Should this be `pump`? Just noticing that it's usually  `pump` elsewhere. Same for the two new tests below.

timcreatedit replied:
Good point, works just as well with `pump`! I think I just copied the three lines from the test right above the ones I added without thinking too much about it. Fixed!

justinmc reviewed (approved):
LGTM :+1:. Thanks for following up after the revert.

I wish we had an automated way to catch this lack of `.up` calls. I was just asking @polina-c about this.

timcreatedit committed:
use pump instead of pumpAndSettle in new tests

dkwingsmt commented:
Ok, I've tested that this will trigger error if the root widget (which supposedly provides the overlay) is take away during the drag.

Reproduction:
1. Apply this PR (or simply access the context during `_DragAvatar.updateDrag`
```dart
  void updateDrag(Offset globalPosition) {
    final RenderBox box = overlayState.context.findRenderObject()! as RenderBox;
    final Offset overlaySpaceOffset = box.globalToLocal(globalPosition);
    print(overlaySpaceOffset - dragStartPoint);
```
2. Run the following app:
```dart
import 'package:flutter/material.dart';

class Main extends StatefulWidget {

  @override
  _MainState createState() => _MainState();
}

class _MainState extends State<Main> {
  bool _show = true;

  @override
  Widget build(BuildContext context) {
    return !_show ? Container() : MaterialApp(
      home: ListView(
        scrollDirection: Axis.horizontal,
        children: <Widget>[
          DragTarget<int>(
            builder: (BuildContext context, List<int?> data, List<dynamic> rejects) {
              return const Text('Target');
            },
            onAcceptWithDetails: (DragTargetDetails<int> _) {
            },
          ),
          Container(width: 400.0),
          Draggable<int>(
            data: 1,
            feedback: Text('H'),
            onDragStarted: () {
              Future<void>.delayed(Duration(seconds: 2)).then((_) {
                setState(() {
                  _show = false;
                });
              });
            },
            childWhenDragging: SizedBox(),
            axis: Axis.horizontal,
            child: Text('H'),
          ),
        ],
      ),
    );
  }
}

void main() {
  runApp(Main());
}
```
3. Drag the `H` and hold for 2 seconds before moving

Throws error:
```

══╡ EXCEPTION CAUGHT BY GESTURE LIBRARY ╞═══════════════════════════════════════════════════════════
The following assertion was thrown while routing a pointer event:
This widget has been unmounted, so the State no longer has a context (and should be considered
defunct).
Consider canceling any active work during "dispose" or using the "mounted" getter to determine if
the State is still active.

When the exception was thrown, this was the stack:
#0      State.context.<anonymous closure> (package:flutter/src/widgets/framework.dart:951:9)
#1      State.context (package:flutter/src/widgets/framework.dart:957:6)
#2      _DragAvatar.updateDrag (package:flutter/src/widgets/drag_target.dart:857:40)
#3      _DragAvatar.update (package:flutter/src/widgets/drag_target.dart:839:5)
#4      MultiDragPointerState._move (package:flutter/src/gestures/multidrag.dart:100:16)
#5      MultiDragGestureRecognizer._handleEvent (package:flutter/src/gestures/multidrag.dart:267:13)
#6      PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:98:12)
#7      PointerRouter._dispatchEventToRoutes.<anonymous closure> (package:flutter/src/gestures/pointer_router.dart:143:9)
#8      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633:13)
#9      PointerRouter._dispatchEventToRoutes (package:flutter/src/gestures/pointer_router.dart:141:18)
#10     PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:127:7)
#11     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:495:19)
#12     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:475:22)
#13     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:450:11)
#14     GestureBinding._handlePointerEventImmediately (package:flutter/src/gestures/binding.dart:420:7)
#15     GestureBinding.handlePointerEvent (package:flutter/src/gestures/binding.dart:383:5)
#16     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:330:7)
#17     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:299:9)
#18     _invoke1 (dart:ui/hooks.dart:328:13)
#19     PlatformDispatcher._dispatchPointerDataPacket (dart:ui/platform_dispatcher.dart:442:7)
#20     _dispatchPointerDataPacket (dart:ui/hooks.dart:262:31)

router: Instance of 'PointerRouter'
route: Closure: (PointerEvent) => void from Function '_handleEvent@171470698':.
event: PointerMoveEvent#dc09b(position: Offset(379.5, 32.1))
════════════════════════════════════════════════════════════════════════════════════════════════════

Another exception was thrown: This widget has been unmounted, so the State no longer has a context (and should be considered defunct). 
```

We should probably fix this error.

dkwingsmt reviewed (changes requested):
See comment

timcreatedit commented:
Thanks for finding this, that's very insightful. As I mentioned on the other PR it didn't quite sit right with me that the test failures were triggered by the changes. I will take a look at this again!

timcreatedit committed:
added failing test for unmounting overlay during drag

timcreatedit committed:
added mounted check to updating drag target position

timcreatedit commented:
@dkwingsmt I added a new test that verifies the behaviour you found when unmounting, and added a mounted check to the context access in `_DragAvatar.updateDrag`. Now the test passes. Do you want me to revert the changes I made to other tests, or do we still want to finish the gestures there? I reckon since unmounting is now covered by a test, we might as well leave it in?

Thanks again for your help!

dkwingsmt reviewed (approved):
LGTM, thanks for working on this!

dkwingsmt commented:
I would prefer removing the `gesture.up`s in existing tests since they are unneeded changes (for the PR) and unneeded lines (for the tests), but I'm not strong on this.

Awesome test for the graceful unmounting, BTW!

justinmc reviewed (approved):
LGTM, but see the analyzer failure.

timcreatedit committed:
removed gesture.up calls

timcreatedit committed:
added explanatory comment to unmount test

timcreatedit commented:
> I would prefer removing the `gesture.up`s in existing tests since they are unneeded changes (for the PR) and unneeded lines (for the tests), but I'm not strong on this.
> 
> 
> 
> Awesome test for the graceful unmounting, BTW!

Removed them all! Thanks :)

timcreatedit commented:
> LGTM, but see the analyzer failure.

@justinmc Ready, it's all green 🌿

timcreatedit commented:
@dkwingsmt @justinmc Just wanted to touch base again, is there anything you still need from me?

dkwingsmt commented:
Nope. Here we go! Thanks for the work!
